components/support/nimbus-fml/src/parser.rs

```diff
@@ -256,109 +278,188 @@ fn collect_channel_defaults(
     Ok(channel_map)
 }
 
-/// Transforms a feature definition with unmerged defaults into a feature
-/// definition with its defaults merged.
-///
-/// # How the algorithm works:
-/// There are two types of defaults:
-/// 1. Field level defaults
-/// 1. Feature level defaults, that are listed by channel
-///
-/// The algorithm gathers the field level defaults first, they are the base
-/// defaults. Then, it gathers the feature level defaults and merges them by
-/// calling [`collect_channel_defaults`]. Finally, it overwrites any common
-/// defaults between the merged feature level defaults and the field level defaults
-///
-/// # Example:
-/// Assume we have the following feature manifest
-/// ```yaml
-///  variables:
-///   positive:
-///   description: This is a positive button
-///   type: Button
-///   default:
-///     {
-///       "label": "Ok then",
-///       "color": "blue"
-///     }
-///  default:
-///      - channel: release
-///      value: {
-///        "positive": {
-///          "color": "green"
-///        }
-///      }
-///      - value: {
-///      "positive": {
-///        "alt-text": "Go Ahead!"
-///      }
-/// }
-/// ```
-///
-/// The result of the algorithm would be a default that looks like:
-/// ```yaml
-/// variables:
-///     positive:
-///     default:
-///     {
-///         "label": "Ok then",
-///         "color": "green",
-///         "alt-text": "Go Ahead!"
-///     }
-///
-/// ```
-///
-/// - The `label` comes from the original field level default
-/// - The `color` comes from the `release` channel feature level default
-/// - The `alt-text` comes from the feature level default with no channel (that applies to all channels)
-///
-/// # Arguments
-/// - `feature_def`: a [`FeatureDef`] representing the feature definition to transform
-/// - `channel`: a [`Option<&String>`] representing the channel to merge back into the field variables
-/// - `supported_channels`: a [`&[String]`] representing the channels that are supported by the manifest
-/// If the `channel` is `None` we default to using the `release` channel
-///
-/// # Returns
-/// Returns a transformed [`FeatureDef`] with its defaults merged
-pub fn merge_feature_defaults(
-    feature_def: FeatureDef,
-    channel: &str,
-    supported_channels: &[String],
-) -> Result<FeatureDef, FMLError> {
-    if !supported_channels.iter().any(|c| c == channel) {
-        return Err(FMLError::InvalidChannelError(channel.into()));
+struct DefaultsMerger<'object> {
```

Comment:
Yeah the merging was and is getting complex 😕 I feel like we're missing a generic struct or trait to help us out - for a future refactor something like: 

```rs
trait Merger {
  type: T1;
  type: T2;
  fn merge(higher_priority: T1, lower_priority: T2) -> serde_json::Value;
}

struct DefaultMerger<T1, T2>;

impl Merger for DefaultMerger<PropDefaults, ObjectDefaults> {
..
}

impl Merger for DefaultMerger<FeatureDefaults, PropDefaults> {
...
}

```

might help reshape this a little

Reply:
Yeah, there's some refactoring to be done here.

We have a `Defaults` trait in `nimbus`, which would be helpful.

I'm relatively sure that there's material work needed to pull out the feature's property defaults and the default blocks into data structures that can be merged, and then back into the feature's property defaults. The merging once in that common data structure is relatively straightfoward.